### PR TITLE
rustdoc: Hide item contents, not items

### LINKED ISSUE #82114: [rustdoc] Have a more nuanced take on what is hidden by default
https://twitter.com/mcclure111/status/1361025145125634061

By default, we:

 - hide attributes
 - hide structs
 - hide unions
 - hide traits
 - hide trait implementation
 - hide implementor list

I feel like traits probably should be shown by default, since everything on the trait is important, especially for implementors.

Furthermore, it might be worth splitting "structs" into "structs with all-public fields" and "structs with private fields" for the same reason (with unit structs being treated as having all public). For the former kind the actual struct declaration is _super_ useful, for the latter kind it's not at all.

cc @rust-lang/rustdoc

### PR DESCRIPTION
**Pull request by Manishearth**
This tweaks rustdoc to hide item contents instead of items, and only when there are too many of them.

This means that users will _always_ see the type parameters, and will _often_ see fields/etc as long as they are small. Traits have some heuristics for hiding only the methods or only the methods and the consts, since the associated types are super important.

I'm happy to play around with the heuristics here; we could potentially make it so that structs/enums/etc are always hidden but traits will try really hard to show type aliases.

This needs a test, but you can see it rendered at https://manishearth.net/sand/doc_render/bar/

<details>

<summary> Code example </summary>

```rust
pub struct PubStruct {
    pub a: usize,
    pub b: usize,
}

pub struct BigPubStruct {
    pub a: usize,
    pub b: usize,
    pub c: usize,
    pub d: usize,
    pub e: usize,
    pub f: usize,
}

pub union BigUnion {
    pub a: usize,
    pub b: usize,
    pub c: usize,
    pub d: usize,
    pub e: usize,
    pub f: usize,
}

pub union Union {
    pub a: usize,
    pub b: usize,
    pub c: usize,
}

pub struct PrivStruct {
    a: usize,
    b: usize,
}

pub enum Enum {
    A, B, C,
    D {
        a: u8,
        b: u8
    }
}

pub enum LargeEnum {
    A, B, C, D, E, F, G, H, I, J
}


pub trait Trait {
    type A;
    #[must_use]
    fn foo();
    fn bar();
}

pub trait GinormousTrait {
    type A;
    type B;
    type C;
    type D;
    type E;
    type F;
    const N: usize = 1;
    #[must_use]
    fn foo();
    fn bar();
}

pub trait HugeTrait {
    type A;
    const M: usize = 1;
    const N: usize = 1;
    const O: usize = 1;
    const P: usize = 1;
    const Q: usize = 1;
    #[must_use]
    fn foo();
    fn bar();
}

pub trait BigTrait {
    type A;
    #[must_use]
    fn foo();
    fn bar();
    fn baz();
    fn quux();
    fn frob();
    fn greeble();
}

#[macro_export]
macro_rules! foo {
    (a) => {a};
}
```

</details>

Fixes https://github.com/rust-lang/rust/issues/82114

### CODE REVIEW

File: src/librustdoc/html/static/rustdoc.css
Comment (camelid) [inline src/librustdoc/html/static/rustdoc.css:1047]: Might be worth a line break here?

File: src/librustdoc/html/render/print_item.rs
Comment (camelid) [inline src/librustdoc/html/render/print_item.rs:463]: Why are you using eager-evaluating `&` instead of `&&` here?

File: src/librustdoc/html/render/print_item.rs
Comment (camelid) [inline src/librustdoc/html/render/print_item.rs:474]: Same as above.

File: src/librustdoc/html/render/print_item.rs
Comment (Manishearth) [reply src/librustdoc/html/render/print_item.rs:463]: oops

**Comment by jyn514** [comment]
> This needs a test, but you can see it rendered at https://manishearth.net/sand/doc_render/bar/

Could you publish a version of the same crate with a nightly rustdoc so I can compare the two?

File: src/librustdoc/html/render/print_item.rs
Comment (jyn514) [inline src/librustdoc/html/render/print_item.rs:1300]:
```suggestion
        .filter(|f| matches!(clean::StructFieldItem(..), *f.kind))
```

File: src/librustdoc/html/render/print_item.rs
Comment (jyn514) [inline src/librustdoc/html/render/print_item.rs:1361]:
```suggestion
                .filter(|f| matches!(clean::StructFieldItem(..) = *f.kind))
```

File: src/librustdoc/html/static/rustdoc.css
Comment (jyn514) [inline src/librustdoc/html/static/rustdoc.css:1764]: What does this do?

File: src/librustdoc/html/render/print_item.rs
Comment (jyn514) [inline src/librustdoc/html/render/print_item.rs]: nit: it might be nice to make this an RAII type that closes the toggle on drop, so you can't forget to close it. But maybe that would be confusing; I'll leave it up to you.

**Review by jyn514 — COMMENTED** [review]
I left a couple comments on the implementation - I'm not super familiar with how the JS frontend works.

If you have time, it would be really helpful to see examples of docs that change before and after for each commit - "enum" and "struct" are clear enough, but I'm not sure what "Stop hiding entire item declarations" does. I know you made a whole site with examples, I'm just not sure which examples have changed.

**Comment by Manishearth** [comment]
Not at home right now, but you can just copy the decls and run rustdoc :)

"Stop hiding item declarations" turns off the thing where we hide _entire_ item decls at once

File: src/librustdoc/html/static/rustdoc.css
Comment (camelid) [reply src/librustdoc/html/static/rustdoc.css:1764]: I don't know the meaning in this particular case, but `>` means select the direct child as opposed to allowing grandchildren to be selected (which is what ` ` does).

File: src/librustdoc/html/render/print_item.rs
Comment (camelid) [reply src/librustdoc/html/render/print_item.rs]: Yeah, I think that would be confusing, and it would be tricky because the `toggle_open` is inside an `if` block, so I think it would be dropped too soon. Maybe a better approach would be to use a `with_toggle` function that takes a closure? It would look like this:

```rust
with_toggle(|| /* do stuff that should be behind a toggle */ );
```

However, that wouldn't work with the conditional toggling determined by the `toggle` variable. Instead we could do something like

```rust
maybe_with_toggle(toggle, || /* do stuff */);
```

where a toggle is only added if `toggle` is true?

**Comment by camelid** [comment]
I think this will need an FCP since it's a fairly significant insta-stable behavioral change. @rustbot label needs-fcp

**Comment by camelid** [comment]
It seems like there's an extra blank line after the toggle that disappears when you click the toggle?

image

image

**Comment by camelid** [comment]
It seems a bit redundant to show the declaration in this case:

image

I think we should maybe hide struct fields by default?

**Comment by camelid** [comment]
(Sorry for all the comments, I'm posting comments as I look through the rendered output.)

Rendering glitch for the attributes, not sure if it was pre-existing or not:

image

**Comment by camelid** [comment]
It doesn't look like you have an example for a small union, just a big one. Could you add an example with a small union?

File: src/librustdoc/html/render/print_item.rs
Comment (Manishearth) [reply src/librustdoc/html/render/print_item.rs]: @jyn514 No, I want the toggle to close at a _very specific_ point.

I thought about adding a closure: I think a problem is that the closure would get _really weird_ for the `trait` stuff. It's cleaner this way, I _started out_ with the other design.

**Comment by Manishearth** [comment]
> It seems like there's an extra blank line after the toggle that disappears when you click the toggle?

Yeah some of this ties in to the styling for toggles being extremely haphazard. I tried fixing this and ended up breaking other things. I can try again.

> I think we should maybe hide struct fields by default?


Maybe! I think we definitely should be applying the complex heuristic for traits, but hidning enums/structs by default seems okay

**Comment by Manishearth** [comment]
> Rendering glitch for the attributes, not sure if it was pre-existing or not:


Ugh no I had to introduce this to make it so that "expand attributes" didn't overlap with "expand methods".

If someone can figure out the CSS for me that would be _really_ helpful; I tried a bunch and couldn't get it to work well. @camelid are you interested in focusing on this part of the work?

File: src/librustdoc/html/static/rustdoc.css
Comment (Manishearth) [reply src/librustdoc/html/static/rustdoc.css:1764]: It used to be a direct descendant selector, but it stopped working because of the new wrapping `docblock` element. Now it will work on all `.toggle-attributes` elements regardless of how far nested they are.

**Comment by Manishearth** [comment]
@jyn514 I added the input code to the PR body

File: src/librustdoc/html/render/print_item.rs
Comment (GuillaumeGomez) [reply src/librustdoc/html/render/print_item.rs]: Maybe at least add an `#[inline]` attribute considering it's a just one line?

File: src/librustdoc/html/render/print_item.rs
Comment (GuillaumeGomez) [inline src/librustdoc/html/render/print_item.rs:1366]: Looking at the code, I'd feel much better if we handled it using a closure like it was suggested. It seems very easy to forget... :-/

File: src/librustdoc/html/static/rustdoc.css
Comment (GuillaumeGomez) [inline src/librustdoc/html/static/rustdoc.css:961]: Questions about this:
 * Does it go above the menu on mobile when the burger button has been pressed?
 * Does it go above the help popup (which appears after "?" has been pressed)?

File: src/librustdoc/html/static/rustdoc.css
Comment (GuillaumeGomez) [inline src/librustdoc/html/static/rustdoc.css:1048]:
```suggestion
	margin-left: 0;
```

File: src/librustdoc/html/render/print_item.rs
Comment (Manishearth) [reply src/librustdoc/html/render/print_item.rs:1366]: I don't think it's possible to have a closure for more complex things like the trait. I could use a drop guard so that it ICEs instead, but that's about it.

Comment (GuillaumeGomez) [reply src/librustdoc/html/render/print_item.rs:1366]: What's preventing to have something like:

```rust
fn wrap_with_toggle<T, E, F: FnOnce() -> Result<T, E>>(f: F) -> Result<T, E> {
   // toggle_open code
  let ret = f()?;
  // toggle_close code
  ret
}
```

? I might miss something obvious here. :-/

Comment (Manishearth) [reply src/librustdoc/html/render/print_item.rs:1366]: Look at the trait code; it calls `toggle_open` at three different points, and `toggle_close` once. 

Again, I _tried_ this, it was too ugly and even more confusing, so I decided not to. I don't think this is too bad. We could add a destructor bomb but honestly we're using this a couple times and shouldn't have to yet. I'd rather get the entire toggle stuff revamped first.

**Comment by camelid** [comment]
> If someone can figure out the CSS for me that would be _really_ helpful; I tried a bunch and couldn't get it to work well. @camelid are you interested in focusing on this part of the work?

I'm afraid my CSS skills are not good enough to be able to do this. If you can't figure out how to do it, I probably won't be able to either.

That said, I might try to look into it a bit in case I magically can figure it out :)

**Comment by Manishearth** [comment]
> If you can't figure out how to do it, I probably won't be able to either.

Nah, it's more like I don't have the time to figure this out. It's _mostly_ just a matter of understanding the existing CSS and working with it, I gave up after trying for a short while and was hoping someone else would be able to look at this.

@jyn514 may also have ideas

**Comment by camelid** [comment]
> Nah, it's more like I don't have the time to figure this out. It's _mostly_ just a matter of understanding the existing CSS and working with it, I gave up after trying for a short while and was hoping someone else would be able to look at this.

I actually just had an idea: what if you just remove the `margin-left` for the attributes? Sure, it's a bit weird because they're not indented, but I think it actually looks fine, and it means you can now read the attributes. It looks like this with the `margin-left` removed:

image

image

**Comment by camelid** [comment]
Hmm, though it looks like there's something off with the indent—it looks like 3-space indent instead of 4...

**Comment by Manishearth** [comment]
> I actually just had an idea: what if you just remove the margin-left for the attributes? Sure, it's a bit weird because they're not indented, but I think it actually looks fine, and it means you can now read the attributes. It looks like this with the margin-left removed:

Then the problem is that they overlap with the other `[-]` from the "show methods" in BigTrait

**Comment by Manishearth** [comment]
We should absolutely be using four spaces for that, yes

**Comment by camelid** [comment]
> We should absolutely be using four spaces for that, yes

Btw, if you are unable to get four spaces working, you can replace the margin-left value of 30px with 4ch. Slightly less hacky and should work, though of course four spaces is ideal.

**Comment by camelid** [comment]
> > I actually just had an idea: what if you just remove the margin-left for the attributes? Sure, it's a bit weird because they're not indented, but I think it actually looks fine, and it means you can now read the attributes. It looks like this with the margin-left removed:
> 
> Then the problem is that they overlap with the other `[-]` from the "show methods" in BigTrait

Ah, that's unfortunate :/

**Comment by Manishearth** [comment]
(To be clear, I don't intend to actively hack on the CSS right now. Maybe later, but folks should feel free to take it up and try improving it)

**Comment by Lokathor** [comment]
Why don't we just always show everything by default? and users can decide to hide things if they want to?

**Comment by Manishearth** [comment]
> Why don't we just always show everything by default? and users can decide to hide things if they want to?

Because it inhibits being able to get the important stuff at a glance. For example, when looking at the trait implementation list users usually want to see a list of all traits, not all the methods for each trait.

In this _specific_ case the wrong decision was made, but overall it's still important for _big_ types to be able to see the actual type docs when you open the page. For example, on `Iterator`; that's a type that should be collapsed.

Note that this PR shows item contents by default, and only starts truncating them if you have too many things inside it (in the current implementation, 5+ fields, but likely will be more like 12 before landing)

**Comment by camelid** [comment]
> Note that this PR shows item contents by default, and only starts hiding them if you have too many things inside it (in the current implementation, 5+ fields, but likely will be more like 12 before landing)

I kind of think we should hide struct fields, enum variants, and union fields by default since all the public fields (and private ones too with `--document-private-items`) are shown, with docs, directly after.

**Comment by Lokathor** [comment]
in the case of very long traits I see your point @Manishearth 

however, for structs, union, and enums i would prefer to simply see the entire declaration, including attributes on the type. If the default allowment is 12 as you suggest that's probably "close enough".

@camelid: the part of the page where each field is then separately listed with its docs is the least useful part of the page. Rust programmers know what Rust code looks like, so the part where they get to see rust code is the good part. The non-code prose stuff after gets in the way between the code declaration of the type (good) and then showing the methods for the type (also good).

**Comment by Manishearth** [comment]
> however, for structs, union, and enums i would prefer to simply see the entire declaration, including attributes on the type. If the default allowment is 12 as you suggest that's probably "close enough".

This will never hide attributes on the type (please look at the example output for this PR if you are going to discuss it further). And yes, my plan for the threshold is 12, The PR picks 5 because 5 is less annoying to test.

> Rust programmers know what Rust code looks like, so the part where they get to see rust code is the good part.

Be careful in presuming what other people need. A common complaint in the pre-hiding days was that really long types would cover up the textual docs. The "Rust code" part and the textual docs are roughly on equal footing when it comes to which is "the good part", with the balance tilting either way based on the specific type.

> the part of the page where each field is then separately listed with its docs is the least useful part of the page

This is debatable. If the fields aren't documented, then maybe, but stuff gets documented for a reason.

Regardless, again, this will only kick in at >12.

**Comment by Manishearth** [comment]
Anyway, I think the overall discussion of the _defaults_ is somewhat off topic for this PR. This is moving the hiding behavior from hiding the entire item (terrible) to hiding just its contents, and we can work on what the default (and the threshold) should be later.

**Comment by jyn514** [comment]
> Not at home right now, but you can just copy the decls and run rustdoc :)

Done: http://50.116.47.176:8000/doc_render/

**Comment by jyn514** [comment]
I like this a lot better. I've seen people ask for a "table of contents" of the documentation a lot and I think this can be something close to that. I didn't realize the links in the declaration were clickable before, that's super useful!

I don't have a strong preference on whether the default should be 5 or 12, if you think 12 I trust your judgement.

I don't have time right this second to look into the CSS issues - cc @Cldfire @jsha maybe one of you would be interested in taking a look at https://github.com/rust-lang/rust/pull/83337#issuecomment-803667259 ?

File: src/librustdoc/html/render/print_item.rs
Comment (jyn514) [reply src/librustdoc/html/render/print_item.rs]: I think this is fine, rustdoc has lots of other performance issues and I wouldn't expect this to make a difference. I didn't realized you'd tried this first, thanks for testing it out :)

**Comment by jsha** [comment]
I'm happy to work on the CSS issues. I'll try to give it a stab later tonight (PDT) or sometime tomorrow. If someone has time before then and wants to work on it don't let me get in the way. :-D

**Comment by camelid** [comment]
> Done: http://50.116.47.176:8000/doc_render/

Hmm, I get a "The connection was reset" error with that page.

**Comment by jyn514** [comment]
> > Done: http://50.116.47.176:8000/doc_render/
> 
> Hmm, I get a "The connection was reset" error with that page.

Apparently `disown` doesn't work the way I thought it did. Should be fixed now, I used tmux instead.

**Comment by Lokathor** [comment]
@Manishearth the example rendering in the first post doesn't appear to *have* type attributes on any of the example types, so it's impossible to tell what your PR does for type attributes ;3

**Comment by Manishearth** [comment]
> @Manishearth the example rendering in the first post doesn't appear to _have_ type attributes on any of the example types, so it's impossible to tell what your PR does for type attributes ;3

It only hides item _contents_, not things outside the `{}`.

File: src/librustdoc/html/static/rustdoc.css
Comment (jsha) [reply src/librustdoc/html/static/rustdoc.css:1764]: Question: Why is there a new wrapping `docblock` element? Can that element be `<div class="type-contents-toggle"  ...>` instead of `<div class="docblock type-contents-toggle"  ...>`?

**Comment by jsha** [comment]
As an update I've made some good progress on layout issues and hope to share a branch soon. Any preferences on how to do that? A pull request against @Manishearth's fork?

**Comment by jsha** [comment]
Posted as a pull request to @Manishearth's fork: https://github.com/Manishearth/rust/pull/1. I wound up including some of the `<details><summary>` work mentioned at https://github.com/rust-lang/rust/pull/83355#issuecomment-803777603. So far I've only applied it to the toggles affected here, but if that gets merged it should set the stage for some followups to replace more of the toggles.

I also removed the toggles for attributes, as discussed above.

**Comment by Manishearth** [comment]
> I wound up including some of the `<details><summary>` work mentioned at [#83355 (comment)](https://github.com/rust-lang/rust/pull/83355#issuecomment-803777603).

Was it necessary to do this? I would rather do these improvements as piecemeal as possible to make them easier to review.

But also, if cleaning up the CSS helps make this work I guess that's fine.

**Comment by Manishearth** [comment]
Actually this is not that hard to review, I got confused a bit. All good, thanks for doing the additional cleanup!

**Comment by Manishearth** [comment]
Alright, this seems quite good, thank you @jsha! I've pushed the rendering at https://manishearth.net/sand/doc_render2/bar/

I think this is mostly ready for review, then, though it needs some tests.

**Comment by Manishearth** [comment]
> I think this will need an FCP since it's a fairly significant insta-stable behavioral change. @rustbot label needs-fcp

@camelid I actually disagree; this isn't "insta stable" since it's totally rollbackable. I think it's _fine_ to do an FCP here, but I don't think "insta-stable" is a good enough reasoning for this since we can undo it whenever we want.

File: src/librustdoc/html/render/print_item.rs
Comment (jsha) [reply src/librustdoc/html/render/print_item.rs:464]: Looks like this is still using `&`, and so is line 475.

File: src/librustdoc/html/static/storage.js
Comment (jsha) [inline src/librustdoc/html/static/storage.js:3]: This unrelated whitespace deletion accidentally snuck in in my branch. Sorry about that. Would be good to revert it to avoid spurious diffs.

**Comment by camelid** [comment]
> I think it's _fine_ to do an FCP here, but I don't think "insta-stable" is a good enough reasoning for this since we can undo it whenever we want.

It's true that we can undo it whenever we want, but it would likely be confusing to users if there's a lot of churn, especially since this is "above the fold".

**Comment by Manishearth** [comment]
Yep. I plan to add tests and change the threshold when I get the chance

**Comment by Manishearth** [comment]
Added a test. r? @jyn514 or @camelid ?

File: src/librustdoc/html/render/print_item.rs
Comment (GuillaumeGomez) [inline src/librustdoc/html/render/print_item.rs:142]: The line is a bit long and has multiple nested levels. What do you think about:

```suggestion
        "<details class=\"rustdoc-toggle type-contents-toggle\">\
            <summary class=\"hideme\">\
                <span>Show {}</span>\
            </summary>",
```

?

File: src/librustdoc/html/render/print_item.rs
Comment (GuillaumeGomez) [reply src/librustdoc/html/render/print_item.rs:464]: The syntax typo is still here. ;)

File: src/librustdoc/html/render/print_item.rs
Comment (GuillaumeGomez) [inline src/librustdoc/html/render/print_item.rs:475]: I'm wondering: is it voluntary?

Comment (Manishearth) [reply src/librustdoc/html/render/print_item.rs:475]: What do you mean?

File: src/librustdoc/html/static/main.js
Comment (GuillaumeGomez) [inline src/librustdoc/html/static/main.js:2700]: I'm very happy to see all this code gone. :D

File: src/librustdoc/html/render/print_item.rs
Comment (Manishearth) [reply src/librustdoc/html/render/print_item.rs:478]: Are you asking if there's a pref? There isn't one right now, I was thinking we'd add one if people _really_ wanted one; but also: 12 is a pretty large threshold so I'm not sure if people really will.

Comment (GuillaumeGomez) [reply src/librustdoc/html/render/print_item.rs:478]: The single `&`. You have it above too so I'm wondering.

File: src/librustdoc/html/static/main.js
Comment (GuillaumeGomez) [inline src/librustdoc/html/static/main.js:2775]: Isn't it still in use for all items with attributes?

File: src/librustdoc/html/static/rustdoc.css
Comment (GuillaumeGomez) [inline src/librustdoc/html/static/rustdoc.css:1769]: Please don't put any color in this file. Add this color rule in the theme instead.

File: src/librustdoc/html/static/rustdoc.css
Comment (GuillaumeGomez) [inline src/librustdoc/html/static/rustdoc.css:1776]: Same as above.

File: src/librustdoc/html/render/print_item.rs
Comment (GuillaumeGomez) [reply src/librustdoc/html/render/print_item.rs:478]: Not talking about the number of items, simply the `&` instead of `&&`. With a diff it might be simpler to see:

```suggestion
            if !toggle && should_hide_fields(required.len() + provided.len()) {
```

File: src/librustdoc/html/static/rustdoc.css
Comment (Manishearth) [reply src/librustdoc/html/static/rustdoc.css]: Done!

File: src/librustdoc/html/static/main.js
Comment (Manishearth) [reply src/librustdoc/html/static/main.js:2775]: Nope, @jsha cleaned it all up :smile:

Comment (Manishearth) [reply src/librustdoc/html/static/main.js:2775]: Nah, @jsha cleaned it up!

File: src/librustdoc/html/render/print_item.rs
Comment (Manishearth) [reply src/librustdoc/html/render/print_item.rs]: Ah! That's a typo, my bad!

File: src/librustdoc/html/render/print_item.rs
Comment (Manishearth) [reply src/librustdoc/html/render/print_item.rs]: Done

Comment (GuillaumeGomez) [reply src/librustdoc/html/render/print_item.rs]: :+1:

File: src/librustdoc/html/render/print_item.rs
Comment (GuillaumeGomez) [reply src/librustdoc/html/render/print_item.rs]: ;)

File: src/librustdoc/html/static/main.js
Comment (GuillaumeGomez) [reply src/librustdoc/html/static/main.js:2775]: Makes me even happier. :D

File: src/librustdoc/html/static/rustdoc.css
Comment (GuillaumeGomez) [reply src/librustdoc/html/static/rustdoc.css]: :+1:

**Review by GuillaumeGomez — APPROVED** [review]

**Comment by Manishearth** [comment]
I added `ignore-tidy-filelength` but it's probably worth splitting main.js up at some point!

**Comment by Manishearth** [comment]
@GuillaumeGomez r=you?

**Comment by Manishearth** [comment]
These tests are no longer necessary since we do not hide attributes anymore. I could make the traits test for the presence of the attribute if necessary

**Comment by camelid** [comment]
> These tests are no longer necessary since we do not hide attributes anymore. I could make the traits test for the presence of the attribute if necessary

Adding a test for the new behavior, even if it *should* never hide, seems like a good idea to me.

**Comment by Manishearth** [comment]
Done

**Comment by jyn514** [comment]
This looks good to me :)

r? @guillaumegomez

**Comment by jyn514** [comment]
@camelid I don't think this needs a full FCP. It's not part of the stability guarantee, it's fairly easy to revert, and I haven't seen anyone object to the new layout.

@rustbot label: -needs-fcp

**Comment by GuillaumeGomez** [comment]
Just finished testing it locally. This is a really nice improvement, thanks everyone!

@bors: r+

**Pull request merged**
